server/service/http/message/query/query_structure.rs

```diff
@@ -55,67 +55,125 @@ impl<'a, Snapshot: ReadableSnapshot> QueryStructureContext<'a, Snapshot> {
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
 pub struct QueryStructureResponse {
-    branches: Vec<QueryStructureBranchResponse>,
+    blocks: Vec<QueryStructureBlockResponse>,
 }
 
 #[derive(Debug, Serialize, Deserialize)]
 #[serde(rename_all = "camelCase")]
-pub struct QueryStructureBranchResponse {
-    edges: Vec<QueryStructureEdgeResponse>,
+pub struct QueryStructureBlockResponse {
+    constraints: Vec<QueryStructureConstraintResponse>,
 }
 
 #[derive(Debug, Serialize, Deserialize)]
-#[serde(rename_all = "camelCase", tag = "kind", content = "param")]
-pub enum QueryStructureEdgeTypeResponse {
-    Isa,
-    Has,
-    Links(QueryStructureVertexResponse),
+#[serde(rename_all = "camelCase", tag = "tag")]
+pub enum QueryStructureConstraint {
+    Isa(QueryConstraintIsaBase),
+    #[serde(rename = "isa!")]
+    IsaExact(QueryConstraintIsaBase),
+    Has {
+        owner: QueryStructureVertexResponse,
+        attribute: QueryStructureVertexResponse,
+        exactness: QueryStructureConstraintExactness,
+    },
+    Links {
+        relation: QueryStructureVertexResponse,
+        player: QueryStructureVertexResponse,
+        role: QueryStructureVertexResponse,
+        exactness: QueryStructureConstraintExactness,
+    },
 
-    Sub,
-    Owns,
-    Relates,
-    Plays,
+    Sub(QueryConstraintSubBase),
+    #[serde(rename = "sub!")]
+    SubExact(QueryConstraintSubBase),
+    Owns {
+        owner: QueryStructureVertexResponse,
+        attribute: QueryStructureVertexResponse,
+        exactness: QueryStructureConstraintExactness,
+    },
+    Relates {
+        relation: QueryStructureVertexResponse,
+        role: QueryStructureVertexResponse,
+        exactness: QueryStructureConstraintExactness,
+    },
+    Plays {
+        player: QueryStructureVertexResponse,
+        role: QueryStructureVertexResponse,
+        exactness: QueryStructureConstraintExactness,
+    },
 
-    IsaExact,
-    SubExact,
+    FunctionCall {
+        name: String,
+        assigned: Vec<QueryStructureVertexResponse>,
+        arguments: Vec<QueryStructureVertexResponse>,
+    },
+    Expression {
+        text: String,
+        assigned: Vec<QueryStructureVertexResponse>,
+        arguments: Vec<QueryStructureVertexResponse>,
+    },
+}
 
-    Assigned(String),
-    Argument(String),
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct QueryConstraintIsaBase {
+    instance: QueryStructureVertexResponse,
+    r#type: QueryStructureVertexResponse,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct QueryConstraintSubBase {
+    subtype: QueryStructureVertexResponse,
+    supertype: QueryStructureVertexResponse,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum QueryStructureConstraintExactness {
+    Exact,
+    Subtypes,
```

Comment:
Can we use 'direct/transitive' for this? Is this even used anymore, if it's rolled into the variant tag??

Reply:
Perhaps we can call it "Transitivity::Direct/Transitive"

Reply:
Removed. Snuck through the `git` ing.